hamt/hamt.go

```diff
@@ -206,26 +206,41 @@ func (ds *Shard) makeShardValue(lnk *ipld.Link) (*Shard, error) {
 
 // Set sets 'name' = nd in the HAMT
 func (ds *Shard) Set(ctx context.Context, name string, nd ipld.Node) error {
+	_, err := ds.SetAndPrevious(ctx, name, nd)
+	return err
+}
+
+// SetAndPrevious sets a link pointing to the passed node as the value under the
+// name key in this Shard or its children. It also returns the previous link
+// under that name key (if any).
+func (ds *Shard) SetAndPrevious(ctx context.Context, name string, node ipld.Node) (*ipld.Link, error) {
```

Comment:
nit: This kind of function is traditionally called `Swap`.

Reply:
Agreed, will be done last in https://github.com/ipfs/go-unixfs/issues/105.